staging/src/k8s.io/apiserver/pkg/apis/apiserverinternal/v1alpha1/types.go

```diff
@@ -0,0 +1,126 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+//  Storage version of a specific resource.
+type StorageVersion struct {
+	metav1.TypeMeta `json:",inline"`
+	// The name is <group>.<resource>.
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+	// Spec is an empty spec. It is here to comply with Kubernetes API style.
+	Spec StorageVersionSpec `json:"spec" protobuf:"bytes,2,opt,name=spec"`
+
+	// API server instances report the version they can decode and the version they
+	// encode objects to when persisting objects in the backend.
+	Status StorageVersionStatus `json:"status" protobuf:"bytes,3,opt,name=status"`
+}
+
+// StorageVersionSpec is an empty spec.
+type StorageVersionSpec struct{}
+
+// API server instances report the versions they can decode and the version they
+// encode objects to when persisting objects in the backend.
+type StorageVersionStatus struct {
+	// The reported versions per API server instance.
+	// +optional
+	// +listType=map
+	// +listMapKey=apiserverID
+	ServerStorageVersions []ServerStorageVersion `json:"serverStorageVersions,omitempty" protobuf:"bytes,1,opt,name=serverStorageVersions"`
```

Comment:
UnderstoodStorageVersions?

Reply:
This field is part of the status, so it's already implied that it's the "understood" version. WDYT?

Reply:
Maybe just "StorageVersions" then?

Reply:
"StorageVersions" sounds good, the only nit is that the top level struct is called "StorageVersion", so it would be StorageVersion.StorageVersionSpec.StorageVersions...

Reply:
That's type names, though? Field names would just be `.spec.storageVersions`, right?

Reply:
Changed the field name and the json label to "StorageVersion". Kept the type name.